src/main/java/com/solopov/urlshorter/web/controllers/UrlShorterController.java

```diff
@@ -0,0 +1,42 @@
+package com.solopov.urlshorter.web.controllers;
+
+import com.solopov.urlshorter.exceptions.LinkNotFoundException;
+import com.solopov.urlshorter.sevices.UrlShorterService;
+import com.solopov.urlshorter.web.dto.UrlGenerateRequestDto;
+import lombok.RequiredArgsConstructor;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import javax.validation.Valid;
+import java.net.URI;
+
+import static org.apache.tomcat.websocket.Constants.MOVED_PERMANENTLY;
+
+@RestController
+@RequestMapping("${com.solopov.urlshorter.service.api.shorten.url}")
+@RequiredArgsConstructor
+public class UrlShorterController {
+
+    private final UrlShorterService service;
+
+    @PostMapping
+    public ResponseEntity<?> generateLink(@Valid @RequestBody UrlGenerateRequestDto dto) {
+        return ResponseEntity
+                .created(URI.create(service.shorten(dto)))
+                .build();
+    }
+
+    @GetMapping("/{shortenUrlId}")
+    public ResponseEntity<?> redirect(@PathVariable String shortenUrlId) {
+        try {
+            return ResponseEntity.status(MOVED_PERMANENTLY)
+                    .header(HttpHeaders.LOCATION, service.fetch(shortenUrlId))
+                    .build();
+
+        } catch (LinkNotFoundException e) {
```

Comment:
the cleaner way would be to map an exception to a response status code using any of the methods mentioned at https://www.baeldung.com/spring-response-status

or have service return Optional instead of throwing the exception

Reply:
Didn't know about this way before. Thanks!

Reply:
upd coommit: added Controller Advice